ENH improve benchopt tests configuration (#942)
benchopt/benchopt: doc/benchmark_workflow/test_benchmark.rst

The test run by ``benchopt test`` will make sure that:

- all datasets have the proper API and can be loaded.

> - the objective have the proper API and can be computed

Author: ```suggestion
- all datasets have a valid API and can be loaded.
- the objective has a valid API and can be computed
```

  with a simple dataset and the result returned by ``get_one_result``.

> - all solvers have the proper API and can be run on a simple

Author: ```suggestion
- all solvers have a valid API and can be run on a simple
```

    The scheduling of the github action run can be changed in
    ``.github/workflows/main.yml``.

Tests reference
~~~~~~~~~~~~~~~

The full set of pytest checks run by ``benchopt test`` (parametrized over

> the benchmark's datasets and solvers where applicable):

Author: ```suggestion
the benchmark's datasets and solvers where applicable) is:
```

- ``test_dataset_class[<dataset>]`` — the dataset class exposes the
  expected public API (``name``, callable ``get_data``).
- ``test_dataset_get_data[<dataset>]`` — each installed dataset's
  ``get_data`` returns a dictionary, as expected by the objective.
- ``test_benchmark_objective[<test_dataset>]`` — the objective instantiates
  on the resolved test dataset and the output of ``Objective.__call__``
  on ``get_one_result`` has the expected schema.
- ``test_benchmark_config_validity`` — the benchmark's ``config.yml``
  only uses valid options.
- ``test_solver_class[<solver>]`` — the solver class exposes the
  expected public API (``name``, ``sampling_strategy``,

>   ``stopping_criterion``, optional ``get_next``).

Author: ```suggestion
  expected public API (``name``, optional ``sampling_strategy``,
  ``stopping_criterion``, or ``get_next``).
```

- ``test_solver_install_api[<solver>]`` — the solver declares a known
  install command (``None``, ``'conda'`` or ``'shell'``).
- ``test_solver_install[<solver>]`` — the solver installs cleanly in a
  fresh conda environment (skipped under ``--skip-install``).
- ``test_solver_stopping_criterion[<solver>-<test_dataset>]`` — the solver's

>   ``stopping_criterion`` is compatible with the objective.

Author: ```suggestion
  ``stopping_criterion`` is compatible with the objective (only useful for
  iterative evaluation benchmark).
```

- ``test_solver_run[<solver>-<test_dataset>]`` — the solver runs on at least
  one configuration of the resolved test dataset. If every variant ends
  up skipped (other than because the solver is not installed),
  ``benchopt test`` synthesises a ``FAILED test_solver_run[<solver>]``
  entry summarising the tested datasets and skip reasons, so coverage

>   gaps surface clearly in CI.

Author: ```suggestion
- ``test_solver_run[<solver>-<test_dataset>]`` — the solver runs on at least
  one configuration of the resolved test datasets.
```

which dataset class is used:

- ``Dataset.test_config`` directly contains the parameters passed to the
  ``Dataset`` class.
- ``Objective.test_config`` directly contains the parameters passed to the
  ``Objective`` class, with an optional ``dataset`` key whose value is a
  dictionary of parameters that override those of the dataset. A special
  ``name`` entry inside that ``dataset`` dictionary selects which dataset
  class is used for testing — and it may be a list of names, in which case
  ``test_benchmark_objective`` is run once per name (useful for objectives
  whose code path differs across datasets, e.g. forecasting vs.
  classification).
- ``Solver.test_config`` directly contains the parameters passed to the
  ``Solver`` class, with optional ``dataset`` and ``objective`` keys whose
  values are dictionaries of parameters that override those of the dataset
  and objective respectively. As with ``Objective.test_config``, a ``name``
  entry inside ``dataset`` (string or list) selects which dataset(s) are
  used to test this particular solver — necessary for benchmarks that ship
  multiple datasets covering different tasks, when a solver only supports
  a subset of them.

> 

Author: ```suggestion
  and objective respectively. A ``name`` parameter in ``dataset`` behaves
  similarly to the one in  ``Objective.test_config``.
```

If none of these is set, the benchmark falls back to a ``Simulated`` test
dataset; the legacy ``Objective.test_dataset_name`` attribute is still
honored for backward compatibility. This default suits most benchmarks, but
real datasets may be too large to be used in tests, and some solvers or
objectives may require specific settings — for instance a specific
regularization level to be fast, or a particular data format only possible
with a given dataset. The ``test_config`` attributes above are the way to
encode these constraints.

> 

Author: ```suggestion
If none of these is set, the benchmark falls back to ``Objective.test_dataset_name``,
which in turn default to ``Simulated``. Configuring such ``test_dataset`` is
necessary to allow testing the benchmark automatically.

```

    class Objective(BaseObjective):
        name = "my objective"
        test_config = {
            'reg': 0.9,
            'dataset': {'name': 'my_data', 'debug': True},
        }

For a multi-task benchmark, a solver that only supports a subset of the
datasets can declare which one(s) to test against:

.. code-block:: python

    class Solver(BaseSolver):
        name = "classifier1"

        test_config = {

>             'dataset': {'name': 'ucr'},  # or ['ucr', 'ucr-multivariate']

Author: ```suggestion
            'dataset': {'name': 'data1'},  # or ['data_a', 'data_b']
```

        }

Note that configuring an empty list of test datasets raises a clear

> ``ValueError`` at collection time.

Author: ```suggestion
Note that configuring an empty list of test datasets raises a
``ValueError`` at collection time.
```
